app/services/cancellation_fee_calculator.rb

```diff
@@ -0,0 +1,28 @@
+class CancellationFeeCalculator
+
+  def initialize(reservation)
+    @reservation = reservation
+  end
+
+  def fee
+    @reservation.blank? ? 0 : calculated_fee
+  end
+
+  private
+
+  def calculated_fee
+    if @calculated_fee.blank?
+
+      # This is wrapped in a transaction and rolled back because
+      # #cancellation_fee can reassign price policies and costs.
+      @reservation.class.transaction do
+        @reservation.canceled_at = Time.current
+        @calculated_fee = @reservation.order_detail.cancellation_fee
```

Comment:
I see that `cancellation_fee` will try to assign a price policy (and possibly pricing) if there isn't already one, but it doesn't look like it would actually hit the database.